### ISSUE TITLE
Consuming from multiple regions with different records per second

### ISSUE THREAD

**Issue by spjegan** [qubole/kinesis-sql#39]
Hi,

First of all, thanks for this wonderful library. It is awesome.

We have a kinesis stream which is spread across 5 different regions. Number of records in each of these regions vary drastically. For instance, one region has 20,000 records/sec while another region has just 500 records/sec.

I am creating these 5 streams and union-ing them together so I could treat it as single stream and perform required operations.

```
  def createStreams(config: Config)(implicit spark: SparkSession): DataFrame = {
    config.getStringList("regions").asScala
      .map(region => createStreamForRegion(region, config))
      .reduce((df1, df2) => df1.union(df2))
  }

  def createStreamForRegion(region: String, config: Config)(implicit spark: SparkSession): DataFrame = {
    spark.readStream
      .format("kinesis")
      .option("streamName", config.getString("stream.name"))
      .option("kinesis.executor.metadata.committer", "hdfs")
      .option("kinesis.executor.metadata.path", s"${config.getString("checkPoints.kinesis")}/$region")
      .option("endpointUrl", s"https://kinesis.$region.amazonaws.com")
      .option("startingPosition", config.getString("stream.startPosition"))
      .option("kinesis.client.describeShardInterval", config.getString("kinesis.describeShardInterval"))
      .option("kinesis.executor.maxRecordPerRead", config.getString("kinesis.maxRecordsPerRead"))
      .option("kinesis.executor.maxFetchRecordsPerShard", config.getString("kinesis.maxRecordsReadPerShard"))
      .option("kinesis.client.numRetries", config.getString("kinesis.numberOfRetries"))
      .option("kinesis.client.retryIntervalMs", config.getString("kinesis.retryIntervalMs"))
      .load
  }
```
With this approach, we see sometimes the KinesisReader lags behind (from current time) on some particular region while other regions are doing just fine.

With Spark Streaming approach, we could configure number of Kinesis Readers per region. So we would configure higher number of readers for regions with more volume and less for other regions. Is there something similar here? How do we make sure that all Kinesis Readers from all regions are up-to-date in consumption and are not lagging behind?

Thanks,
Jegan

**Comment by itsvikramagr** [comment]
@spjegan -  This is an interesting use-case. I haven't thought about it while implementing this library. In fact, we have an open issue https://github.com/qubole/kinesis-sql/issues/33 to handle multiple streams. Multiple streams across regions add more color to this problem. We may have to add logic to create multiple tasks per stream shards. 

For now, can you try increasing the value of `kinesis.executor.maxFetchRecordsPerShard` to say a higher value (say 5 sec or higher value as per your requirement)? Each task will be blocked for 5s or until it reads `kinesis.executor.maxRecordPerRead` (10000) records. This will make sure each micro-batch spends enough time on reading records.

**Comment by spjegan** [comment]
@itsvikramagr Thanks for the response.

I guess you meant `kinesis.executor.maxFetchTimeInMs` property. Correct me if I am wrong. I will try with this property and update here.

If this works out, we are planning to use this library in production very soon (probably in 2-3 weeks). Do you have any design in mind to support this use-case? Is there anything I could help out with?

**Comment by itsvikramagr** [comment]
@spjegan - I have edited the response with the correct property name.

For optimal utilization for resources for streams across regions, we need to understand the reason for the lag. Can you find out if the lag is due to slow response of AWS API for a particular region or is it due to a higher ingestion rate in one of the streams?

**Comment by spjegan** [comment]
@itsvikramagr I do not really see any any issues in response times with Kinesis APIs. They all are under 100 ms on average across all regions. And the ingestion rate is also normal. I checked these using the Kinesis metrics. I also trimmed the application to a bare minimum just to rule out issues with application. As a mater of fact, we have an old pipeline, based on Spark Streaming, that runs fine without any lag. Hope this answers your questions. 

I also tried setting `kinesis.executor.maxFetchTimeInMs` to 20 seconds, but without any luck. Please let me know if you want me to check anything else.

**Comment by itsvikramagr** [comment]
Old spark streaming has receiver based approach which had its own limitations.  Here every micro-batch consists of reading from Kinesis, processing it and then writing to a sink. There are no separate receivers to only read from Kinesis.

I would need to look at the logs (executors and driver) to understand what is happening.  If you share those to me at vagrawal @ qubole.com, I can spend some time on it.

**Comment by itsvikramagr** [comment]
We can start with only `kinesis-sql` logs. If required I can ask for more

One task per shard is created. If there any enough workers to work all on all tasks, they can work parallelly. Otherwise, some tasks will be picked up once prev task is completed. But that might increase your overall micro-batch execution time but should not impact the lag. We have to look into logs to find more.

Qubole has open-sourced [spark-lens ](http://sparklens.qubole.net/) for tuning batch jobs and some of us are now working on tuning streaming jobs.

**Comment by spjegan** [comment]
@itsvikramagr Sorry for the delay. I will try to generate the logs for `kinesis-sql` and share it with you by tomorrow (PST).
And thanks for sharing `spark-lens` project. Looks really interesting. Would definitely give it a try.

**Comment by spjegan** [comment]
@itsvikramagr Upon further investigation I found out that the configured value for `kinesis.executor.maxFetchTimeInMs` was NOT taken into account by Kinesis Reader. It was always using the default value (1000) because of an issue (#36) in parsing configurations. This was happening even for few other configurations. After including the fix from #36, it started running fine. I have not seen any significant delay so far as earlier. The delay is always within a minute which is acceptable to us. So your initial suggestion of using `maxFetchTimeInMs` works. Thanks a lot for your suggestion.

Will keep this open for few more days and close it once I verify it completely.
